packages/context/src/reflect.ts

```diff
@@ -3,18 +3,153 @@
+ */
+export class NamespacedReflect {
+  /**
+   * @param namespace : namespace to bind this reflect context
+   */
+  constructor(private namespace?: string) {}
+
+  private getMetadataKey(metadataKey: string): string {
+    // prefix namespace, if provided, to the metadata key
+    return this.namespace? this.namespace + ':' + metadataKey : metadataKey;
+  }
+
+  /**
+   * define metadata for a target class or it's property/method
+   */
+  defineMetadata(metadataKey: string, metadataValue: any, target: Object, propertyKey?: string | symbol) {
+    metadataKey = this.getMetadataKey(metadataKey);
+    if (propertyKey) {
+      Reflect.defineMetadata(metadataKey, metadataValue, target, propertyKey);
+    }
+    else {
+      Reflect.defineMetadata(metadataKey, metadataValue, target);
+    }
+  }
+
+  /**
+   * lookup metadata from a target object and its prototype chain
+   */
+  getMetadata(metadataKey: string, target: Object, propertyKey?: string | symbol): any {
+    metadataKey = this.getMetadataKey(metadataKey);
+    if (propertyKey) {
+      return Reflect.getMetadata(metadataKey, target, propertyKey);
+    }
+    return Reflect.getMetadata(metadataKey, target);
+  }
+
+  /**
+   * get own metadata for a target object or it's property/method
+   */
+  getOwnMetadata(metadataKey: string, target: Object, propertyKey?: string | symbol): any {
+    metadataKey = this.getMetadataKey(metadataKey);
+    if (propertyKey) {
+      return Reflect.getOwnMetadata(metadataKey, target, propertyKey);
+    }
+    return Reflect.getOwnMetadata(metadataKey, target);
+  }
+
+  /**
+   * Check if the target has corresponding metadata
+   * @param metadataKey Key
+   * @param target Target
+   * @param propertyKey Optional property key
+   */
+  hasMetadata(metadataKey: string, target: Object, propertyKey?: string | symbol): boolean {
+    metadataKey = this.getMetadataKey(metadataKey);
+    if (propertyKey) {
+      return Reflect.hasMetadata(metadataKey, target, propertyKey);
+    }
+    return Reflect.hasMetadata(metadataKey, target);
+  }
+
+  hasOwnMetadata(metadataKey: string, target: Object, propertyKey?: string | symbol): boolean {
+    metadataKey = this.getMetadataKey(metadataKey);
+    if (propertyKey) {
+      return Reflect.hasOwnMetadata(metadataKey, target, propertyKey);
+    }
+    return Reflect.hasOwnMetadata(metadataKey, target);
+  }
+
+  deleteMetadata(metadataKey: string, target: Object, propertyKey?: string | symbol): boolean {
+    metadataKey = this.getMetadataKey(metadataKey);
+    if (propertyKey) {
+      return Reflect.deleteMetadata(metadataKey, target, propertyKey);
+    }
+    return Reflect.deleteMetadata(metadataKey, target);
+  }
+
+  getMetadataKeys(target: Object, propertyKey?: string | symbol): string[] {
+    let keys: string[];
+    if (propertyKey) {
+      keys = Reflect.getMetadataKeys(target, propertyKey);
+    } else {
+      keys = Reflect.getMetadataKeys(target);
+    }
+    const metaKeys = [];
```

Comment:
nitpick, can we change this to `let metaKeys;` and assign empty array [] only on required condition